### ISSUE TITLE
Alias for nushell

### ISSUE THREAD

**Issue by fellnerse** [nvbn/thefuck#1254]
Hey I'm currently trying out [nushell](https://www.nushell.sh/) and wondering if you would like to add the alias to the `thefuck -a` command:

for me this works: `alias fuck = $"(TF_ALIAS=fuck PYTHONIOENCODING=utf-8 thefuck $'(history | last 1)')"`

Thanks in advance.

**Comment by scorphus** [comment]
I didn't try nushell so I may be wrong. But seems like nushell/nushell/issues/2812 forbids thefuck from working with nushell.

**Comment by fellnerse** [comment]
Yes sourcing works differently in nushell and I had trouble to get it to work but the setup for thefuck seems to be way easier. So if you print the line I wrote above when executing `thefuck -a` and the user puts it in the config this will work. I just thought I should share my input so the next person that tries this does not need to go through the same hassle.

**Comment by scorphus** [comment]
I still wonder how the fixed command is relayed to nushell and then executed.

**Comment by fellnerse** [comment]
yes you are right, I forgot that I piped it into bash. But in the mean time I found a better solution/'hack' that is also used in other tools supporting nushell:
`alias fuck = $"(TF_ALIAS=fuck PYTHONIOENCODING=utf-8 thefuck $'(history | last 1)' | save /tmp/fuck.nu; nu /tmp/fuck.nu)"`
Currently saving and then executing a file is the only way for running a string as a command. For that reason I implemented something similar to `psub` from fishshell which basically does the [same](https://github.com/fish-shell/fish-shell/blob/master/share/functions/psub.fish)

**Comment by scorphus** [comment]
I'd say you're set to start adding support to nushell in thefuck. What would you say? :slightly_smiling_face:

**Comment by fellnerse** [comment]
Sure, I'd love to, how do I do that?

**Comment by scorphus** [comment]
Awesome! You can start by creating a new class for nushell, extending from `Generic`, in the same manner as https://github.com/nvbn/thefuck/blob/master/thefuck/shells/fish.py, overriding `Generic` with nushell-specific functionality.

Then add it to:

https://github.com/nvbn/thefuck/blob/0f4a523dc4cab990641131a7aac90b67bf7fb5fc/thefuck/shells/__init__.py#L14-L20

**Comment by shersch** [comment]
hey now!  sorry to raise the dead here, but I am now digging into nushell and would love to bring thefuck with me over to it.  I added this to my config as @fellnerse suggested:
`alias fuck = $"(TF_ALIAS=fuck PYTHONIOENCODING=utf-8 thefuck $'(history | last 1)' | save -f /tmp/fuck.nu; nu /tmp/fuck.nu)"`

but it still isn't working.  I entered 'lss' and got an error as expected:
`Error: nu::shell::external_command (link)

  × External command failed
   ╭─[entry #11:1:1]
 1 │ lss
   · ─┬─
   ·  ╰── did you mean 'ls'?
   ╰────
  help: No such file or directory (os error 2)`

then 'fuck', and got this:
`command lss, index: 267}] [enter/↑/↓/ctrl+c]`

when I hit enter, I get this
`Error: nu::parser::unbalanced_delimiter (link)

  × Unbalanced delimiter.
   ╭─[/private/tmp/fuck.nu:1:1]
 1 │ command lss, index: 267}]
   ·                        ▲
   ·                        ╰── unbalanced { and }
   ╰────`

any help would be appreciated.  sorry in advance for the awful formatting i'm expecting this to have when i smash that green button.

**Comment by scorphus** [comment]
Seems like it's necessary to parse the result of `$'(history | last 1)'`.

**Comment by shersch** [comment]
thanks for the response!  I'm pretty new to all of this, so forgive me if I'm doing something wrong.  I gave this a try:
`alias fuck = $"(TF_ALIAS=fuck PYTHONIOENCODING=utf-8 thefuck parse $'(history | last 1)' | save -f /tmp/fuck.nu; nu /tmp/fuck.nu)" `

It does let me choose a command, kind of.  It always shows "paste" as the first option regardless of what incorrect command I entered before.  After hitting enter, I get an error that says "cannot convert record".  But I'm also noticing that it's auto-completing to "fufuck" for some reason as well, and "fuck" returns this message:
`Seems like fuck alias isn't configured!
More details - https://github.com/nvbn/thefuck#manual-installation`

<img width="1129" alt="CleanShot 2023-03-15 at 06 56 02@2x" src="https://user-images.githubusercontent.com/10481382/225160213-164a3a0a-05ea-4105-a22c-b0f2c75aacf1.png">

**Comment by scorphus** [comment]
> Awesome! You can start by creating a new class for nushell, extending from `Generic`, in the same manner as https://github.com/nvbn/thefuck/blob/master/thefuck/shells/fish.py, overriding `Generic` with nushell-specific functionality.
> 
> Then add it to:
> 
> https://github.com/nvbn/thefuck/blob/0f4a523dc4cab990641131a7aac90b67bf7fb5fc/thefuck/shells/__init__.py#L14-L20

The parsing should take place in the aforementioned, new class.

**Comment by shersch** [comment]
I thought I'd take a stab at this, but I'm very much out of my league.  Python is all fine to me, but I think I know way too little about Nushell at this point to be able to do this.

**Comment by zhanghua000** [comment]
Maybe this one should be fine:
```nu
alias fuck = with-env {TF_ALIAS: "fuck", PYTHONIOENCODING: "utf-8"} {
    thefuck (history | last 1 | get command.0)
}
```

**Comment by ZerdoX-x** [comment]
What about instant mode?

**Comment by Skyppex** [comment]
I managed to get it working just now and thought i'd share my findings:
I'm on windows using scoop as my package manager. I had some issues when installing thefuck using scoop, but i think after a restart it seemed to start working. For anyone wondering, i install thefuck from anderlli0053's scoop bucket called DEV-tools.
Scoop was managing my python installation which is on a version where the `imp` module has been dropped in favor of `importlib`. All i had to do was change some of the source code. Where the command was showing an error there was a usage of the `load_source` function imported from `imp`. I replaced with my own defenition:
```py
import importlib.util
import importlib.machinery

def load_source(modname, filename):
    loader = importlib.machinery.SourceFileLoader(modname, filename)
    spec = importlib.util.spec_from_file_location(modname, filename, loader=loader)
    module = importlib.util.module_from_spec(spec)
    # The module is always executed and not cached in sys.modules.
    # Uncomment the following line to cache the module.
    # sys.modules[module.__name__] = module
    loader.exec_module(module)
    return module
```
Source: https://stackoverflow.com/questions/76694726/replacing-imp-load-source-in-python-3-12
I did this in two places, but i'm not sure if its necessary. It errored once when i called `fuck` and again when i called `thefuck` from the command line so i just replaced both to be safe.
After that all i had to do was paste in the `alias`.

For me an `alias` doesn't support expressions or something like that so i made it a `def` instead. Here it is:
```nu
def fuck [
    --yes(-y),
    --yeah,
    --hard,
    --repeat(-r),
    --debug(-d)
    --enable-experimental-instant-mode,
    --shell-logger(-l)
    --help(-h),
    --version(-v),
] {
    if $help { thefuck -h; return }
    if $version { thefuck -v; return }

    mut options = []

    if $yes or $yeah or $hard { $options = ($options | append "-y") }
    if $repeat { $options = ($options | append "-r") }
    if $debug { $options = ($options | append "-d") }
    if $enable_experimental_instant_mode { $options = ($options | append "--enable-experimental-instant-mode") }
    if $shell_logger { $options = ($options | append "-l") }

    let options = ($options | str join " ")
    let hist = (history | last 1) | get command.0
    let fuck = $"(TF_ALIAS=fuck PYTHONIOENCODING=utf-8 thefuck $options $hist)"
    nu -e $"($fuck | str replace -a "\r" "" | str replace -a "\n" "")"
}
```
Its a lot more code i admit, bit the other examples didn't work for me so here we are :D
Here i'm running `nu` with the `-e` flag and just passing the command as a string. This way if you do a command such as `cd` it still works. Without the `-e` flag, nu runs the command in a seperate process and forgets about it. With `-e` it runs it in a seperate process and then changes the current one to that new process kinda...

This also avoid saving to a temp file :D.

I tested several things here. One thing to note is the `--shell-logger` flag can be removed if using windows since it only works on Linux or Mac, but i included it here anyway.
Tested with multi-line commands, seems to work fine actually, although `thefuck` doesn't know about the internal nu commands making it worse than it needs to be. I ran
```nu
loop {
    print hello
    sleep 1sec
}
```
and it changed it to
```nu
scoop {
    print hello
    sleep 1sec
}
```
which makes sense. `scoop` is an executable in my PATH, but the nu commands arent, so it doesn't know about them and thus, doesn't use them for fixing commands.
Hope this helps someone who just want to hack this together like i did :D

**Comment by PeterhrAda** [comment]
Based on the previous answers, I made a version that works for me:

```
def f [] {
    with-env {TF_ALIAS: "f", PYTHONIOENCODING: "utf-8"}  {
	    let cmd = thefuck (history | last 1 | get command.0)
	    if ( $cmd | is-not-empty ) {
		    nu -c $cmd
	    }
    }
}
```
